akka-typed/src/main/scala/akka/typed/ScalaProcess.scala

```diff
@@ -0,0 +1,538 @@
+            }
+        step(n, Nil)
+      }
+    }
+    private object nextLoop extends NextLoop[Nothing]
+
+    def apply[T]: Next[T] = next.asInstanceOf[Next[T]]
+
+    trait Next[T] {
+      def apply[U](body: OpDSL { type Self = T } ⇒ Operation[T, U]): Operation[T, U] =
+        unit[T, OpDSL { type Self = T }].flatMap(body)
+    }
+    private object next extends Next[Nothing]
+
+    trait NextStep[T] {
+      def apply[U](mailboxCapacity: Int, body: OpDSL { type Self = T } ⇒ Operation[T, U])(implicit opDSL: OpDSL): Operation[opDSL.Self, U] =
+        Call(Process("nextStep", Duration.Inf, mailboxCapacity, body(null)))
+    }
+    private[typed] object nextStep extends NextStep[Nothing]
+  }
+
+  /*
+   * Terminology:
+   *
+   *  - a Process has a 1:1 relationship with an ActorRef
+   *  - an Operation is a step that a Process takes and that produces a value
+   *  - Processes are concurrent, but not distributed: all failures stop the entire Actor
+   *  - each Process has its own identity (due to ActorRef), and the Actor has its own
+   *    identity (an ActorRef[ActorCmd[_]]); processSelf is the Process’ identity, actorSelf is the Actor’s
+   *  - timeout means failure
+   *  - every Actor has a KV store for state
+   *      - querying by key in TypedMultiMap (using a single element per slot)
+   *      - updating is an Operation taking an Event and an implicit (Event, State) => State
+   *      - persistence can then be plugged in transparently
+   *      - recovery means acquiring state initially (which might trigger internal replay)
+   */
+
+  /**
+   * Helper to make `Operation.map` behave like `flatMap` when needed.
+   */
+  trait MapAdapter[Self, Out, Mapped] {
+    def lift[O](f: O ⇒ Out): O ⇒ Operation[Self, Mapped]
+  }
+  /**
+   * Helper to make `Operation.map` behave like `flatMap` when needed.
+   */
+  object MapAdapter extends MapAdapterLow {
+    implicit def mapAdapterOperation[Self, M]: MapAdapter[Self, Operation[Self, M], M] =
+      new MapAdapter[Self, Operation[Self, M], M] {
+        override def lift[O](f: O ⇒ Operation[Self, M]): O ⇒ Operation[Self, M] = f
+      }
+  }
+  /**
+   * Helper to make `Operation.map` behave like `flatMap` when needed.
+   */
+  trait MapAdapterLow {
+    implicit def mapAdapterAny[Self, Out]: MapAdapter[Self, Out, Out] =
+      new MapAdapter[Self, Out, Out] {
+        override def lift[O](f: O ⇒ Out): O ⇒ Operation[Self, Out] = o ⇒ Return(f(o))
+      }
+  }
+
+  /**
+   * A Process runs the given operation steps in a context that provides the
+   * needed [[ActorRef]] of type `S` as the self-reference. Every process is
+   * allotted a maximum lifetime after which it is canceled; you may set this
+   * to `Duration.Inf` for a server process.
+   */
+  final case class Process[S, +Out](name: String, timeout: Duration, mailboxCapacity: Int, operation: Operation[S, Out]) {
+    if (name != "") a.ActorPath.validatePathElement(name)
```

Comment:
because this is not the full name: keeping those unique would be too much of a burden, and it would not be composable; this name will eventually be appended to the generated `$!a` etc. (but that is not yet implemented)